### PR TITLE
AJ-1284: More Prefactoring for supporting `UPDATING` on Data tab

### PR DESCRIPTION
**Pull request by jladieu**
## Summary of changes:
* Add test coverage to verify polling behavior, which helped discover a bug which is that polling didn't _stop_ after reaching various `ERROR` states.
* Improve polling logic to respect `ERROR` states as terminal.

### Testing strategy
- Added new and continued passing existing unit tests.

### CODE REVIEW

File: src/data/Data.js [hunk 1]
```diff
@@ -769,22 +768,22 @@ export const WorkspaceData = _.flow(
 
     const loadWdsData = useCallback(async () => {
       // Try to load the proxy URL
-      if (!wdsProxyUrl || wdsProxyUrl.status !== 'Ready') {
+      if (!wdsProxyUrl || !['Ready', 'Error'].includes(wdsProxyUrl.status)) {
```
Comment (jladieu) [inline src/data/Data.js:771]: This is getting a bit grosser before it gets better.  I hope to still tighten this logic up a bit, but I thought it prudent not to tie up a fix to the polling behavior behind the cleanup.

File: src/data/Data.test.ts [hunk 2]
```diff
@@ -190,21 +199,120 @@ describe('WorkspaceData', () => {
...
+    expect(mockGetSchema).not.toHaveBeenCalled(); // don't fetch schema yet
+    expect(screen.queryByText(/Select a data type/)).toBeNull();
+    expect(screen.getByText(/Preparing your data tables/)).toBeVisible();
+
+    // Act
+    await act(async () => {
+      jest.advanceTimersByTime(30000);
+    });
+
+    // Assert
```
Comment (jladieu) [inline src/data/Data.test.ts:310]: Does AAAAAAA violate the spirit of AAA if it's testing polling? :D

Comment (nawatts) [reply src/data/Data.test.ts:310]: _I_ think it's fine. I prefer to write tests this way vs storing intermediate results in order to have one act and one assert section.

**Review by nawatts — COMMENTED** [review]
Since this is getting into multiple rounds of refactoring anyway, would it be worth going all the way and extracting the polling for the WDS app into a hook?

Something like `usePolledWdsApp`? It could make use of [`usePollingEffect`](https://github.com/DataBiosphere/terra-ui/blob/d9e11ff93756f0b7296d6e5038550fd225a7ebbb/src/libs/react-utils.ts#L146) and splitting it out into a separate module would allow writing it in TS.

**Comment by jladieu** [comment]
> Since this is getting into multiple rounds of refactoring anyway, would it be worth going all the way and extracting the polling for the WDS app into a hook?
> 
> Something like `usePolledWdsApp`? It could make use of [`usePollingEffect`](https://github.com/DataBiosphere/terra-ui/blob/d9e11ff93756f0b7296d6e5038550fd225a7ebbb/src/libs/react-utils.ts#L146) and splitting it out into a separate module would allow writing it in TS.

Yes!  I started looking at (and being confused by `usePollingEffect` and really destabilized the code in the process!  Doing it with TS' support is maybe a better play.

That said, I do think this is a functional improvement in that it fixes a defect and improves test coverage, so if there's nothing terribly wrong with this PR I'd still like to merge if possible in the spirit of shipping the smallest possible batch of changes at once.

**Review by nawatts — APPROVED** [review]

**Review by davidangb — APPROVED** [review]

**Pull request merged**